### PR TITLE
[IMP] website_slides: improve the content player look

### PR DESCRIPTION
**Pull request by dht-odoo**
PURPOSE

Improve the "Content Player" look and share dialog modal look.

SPECIFICATION
Currently:
- there was a strange flicker of 'Add a review' and 'Write a review'
  while loading the course player because we have different strings
  in xml and js due to which while loading it appears like flickering
  of 'Add a review' and 'Write a Review'.
- there were very few options for sharing course in website_slides
  compare to the website snippet of share.
- we are displaying share icon with string 'Share' and for edit review
  we have string 'Modify your review'.
- full url was not visible for sharing.
- we cannot enter multiple emails in Share by email field as it was an input
  field with type email.

To Be:
- there will be no flickering as we have modify the string 'Write a review'
  to 'Add a review'.
- we have added 'whatsapp' and 'pinterest' options for sharing.
- we have modified the view of share dialog box by rearranging the fields
  inside it. And also full url of sharing will be visible.
- we have removed the string 'Share' and modified the string 'Modify your
  review' to 'Edit review'.
- we have converted the input field of Share by email type email to text, so
  we can enter multiple emails to share and also added vaidation in jsfor this
  input field.

Task-2627364

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/static/src/js/slides_embed.js [hunk 1]
```diff
@@ -226,8 +226,15 @@ $(function () {
...
+            const inputValues = input.val().split(",");
+            let areEmailsValid = true;
             var slideID = widget.find('button').data('slide-id');
-            if (input.val() && input[0].checkValidity()) {
+            inputValues.forEach(email => {
+                if (!email.match(/.+@.+/)) {
+                    areEmailsValid = false;
+                }
+            });
+            if (input.val() && areEmailsValid) {
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/slides_embed.js:237]: This code is copy / paste 3 times in this commit, isn't it better to do the verification in `/slides/slide/send_share_email` and return an error if one of the email is not valid ?

With this you can also use the existing email regex in `tool/mail.py`, again we will remove redundancy :)

What do you think ?

File: addons/website_slides/static/src/js/slides_share.js [hunk 2]
```diff
@@ -29,9 +36,10 @@ var ShareMail = publicWidget.Widget.extend({
                     email: input.val(),
                 },
             }).then(function () {
-                self.$el.html($('<div class="alert alert-info" role="alert">' + _t('<strong>Thank you!</strong> Mail has been sent.') + '</div>'));
+                self.$el.html($('<div class="alert alert-info" role="alert">' + _t('<strong>Sharing is caring!</strong> Email(s) sent.') + '</div>'));
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/slides_share.js:39]: Isn't it possible to put this message in the dom with a class `d-none` and remove the class to show the error ?

So you will not need to call `.html` and it will be easier to see where the message is inserted, what do you think ?

**Review by std-odoo — COMMENTED** [review]
Hi :)

I added some comments, can you have a look ? :) 

Thank you

File: addons/website_slides/static/src/js/slides_embed.js [hunk 3]
```diff
@@ -236,13 +243,14 @@ $(function () {
...
                     success: function () {
-                        widget.html($('<div class="alert alert-info" role="alert"><strong>Thank you!</strong> Mail has been sent.</div>'));
+                        widget.html($('<div class="alert alert-info" role="alert"><strong>Sharing is caring!</strong> Email(s) sent.</div>'));
                     },
                     error: function (data) {
                         console.error("ERROR ", data);
                     },
                 });
             } else {
+                window.alert('Please enter valid email(s)');
```
[the remark this answers is not in the document]
Comment (dht-odoo) [reply addons/website_slides/static/src/js/slides_embed.js:253]: This is for embed code and I couldn't check from where in the embedded content we can share the email
that's why I used alert. Can you suggest how can we check?
Thanks in advance!

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_slides/static/src/js/slides_embed.js:253]: like above, you can show the error message with `widget.html` e.g.

File: addons/portal_rating/static/src/js/portal_rating_composer.js [hunk 4]
```diff
@@ -94,7 +94,7 @@ const RatingPopupComposer = publicWidget.Widget.extend({
             // Change the text of the button
             this.$('.o_rating_popup_composer_text').text(
                 this.options.default_message_id ?
-                _t('Modify your review') : _t('Add a review')
+                _t('Edit review') : _t('Add a review')
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/portal_rating/static/src/js/portal_rating_composer.js:97]: not solved, you still have 2 redundant section in the second commit message
still not solved, you have 2 section redundant in the second commit message

> sharing content with multiple emails was not supported

> we can enter multiple emails to share

File: addons/website_slides/controllers/main.py [hunk 5]
```diff
@@ -841,6 +842,10 @@ def slide_preview(self, slide_id):
 
     @http.route(['/slides/slide/send_share_email'], type='json', auth='user', website=True)
     def slide_send_share_email(self, slide_id, email, fullscreen=False):
+        email_ids = email.split(",")
```
Comment (std-odoo) [inline addons/website_slides/controllers/main.py:845]: (Picky remark: if a file use single quote, use single quote for consistency)
```suggestion
        email_ids = email.split(',')
```

File: addons/website_slides/controllers/main.py [hunk 6]
```diff
@@ -841,6 +842,10 @@ def slide_preview(self, slide_id):
...
+        for email_id in email_ids:
```
Comment (std-odoo) [inline addons/website_slides/controllers/main.py:846]: Why `email_id` ? it just a email isn't it ?
```suggestion
        for email in emails:
```

File: addons/website_slides/static/src/js/slides_embed.js [hunk 7]
```diff
@@ -227,22 +227,33 @@ $(function () {
...
                 $.ajax({
                     type: "POST",
                     dataType: 'json',
                     url: '/slides/slide/send_share_email',
                     contentType: "application/json; charset=utf-8",
                     data: JSON.stringify({'jsonrpc': "2.0", 'method': "call", "params": {'slide_id': slideID, 'email': input.val()}}),
-                    success: function () {
-                        widget.html($('<div class="alert alert-info" role="alert"><strong>Thank you!</strong> Mail has been sent.</div>'));
+                    success: function (action) {
+                        // We are uusing action.result here because we get other two elements 'jsonrpc' and 'id' other than 'result'.
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/slides_embed.js:239]: comment not really useful

**Review by std-odoo — COMMENTED** [review]
It's a lot better now :) I still have a few comment, can you have a look ?

File: addons/website_slides/controllers/main.py (hunk 6 above)
Comment (dht-odoo) [reply addons/website_slides/controllers/main.py]: Here we have conflict for name 'email' with function argument name. So, I have used mail and mails here.
    for mail in mails:

Comment (std-odoo) [reply addons/website_slides/controllers/main.py]: the function argument should be `emails` it contain many emails :)

Comment (dht-odoo) [reply addons/website_slides/controllers/main.py]: Changes done.

File: addons/website_slides/controllers/main.py [hunk 8]
```diff
@@ -909,10 +910,20 @@ def slide_preview(self, slide_id):
...
-        result = slide._send_share_email(email, fullscreen)
-        return result
+    def slide_send_share_email(self, emails, res_id, fullscreen=False):
+        for email in emails.split(','):
+            if not single_email_re.match(email.strip()):
+                return False
+        slide = request.env['slide.slide'].browse(int(res_id))
+        return slide._send_share_email(emails, fullscreen)
+
+    @http.route(['/slides/channel/send_share_email'], type='json', auth='user', website=True)
```
Comment (std-odoo) [inline addons/website_slides/controllers/main.py:920]: can you merge this route with the previous one ?

you can even give the model in the route

File: addons/website_slides/data/mail_template_data.xml [hunk 9]
```diff
@@ -33,6 +33,36 @@
...
+                        Hello<br/><br/>
+                        <t t-out="user.name or ''">Mitchell Admin</t> shared the <strong t-out="object.name or ''">document</strong> with you!
+                        <div style="margin: 16px 8px 16px 8px; text-align: center;">
+                            <a t-att-href="object.website_url">
+                                <img t-att-alt="object.name" t-attf-src="{{ ctx.get('base_url') }}/web/image/slide.channel/{{ object.id }}/image_1024" style="height:auto; width:150px; margin: 16px;"/>
+                            </a>
+                        </div>
+                        <div style="padding: 16px 8px 16px 8px; text-align: center;">
+                            <a t-att-href="object.website_url"
+                                style="background-color: #875a7b; padding: 8px 16px 8px 16px; text-decoration: none; color: #fff; border-radius: 5px;">View <strong t-out="object.name or ''">Document</strong></a>
```
Comment (std-odoo) [inline addons/website_slides/data/mail_template_data.xml:54]: can you split this line ?

File: addons/website_slides/data/mail_template_data.xml [hunk 10]
```diff
@@ -33,6 +33,36 @@
...
+            <field name="email_from">{{ user.email_formatted }}</field>
+            <field name="email_to">{{ ctx.get('email', '') }}</field>
+            <field name="body_html" type="html">
+                <div style="margin: 0px; padding: 0px;">
+                    <p style="margin: 0px; padding: 0px; font-size: 13px;">
+                        Hello<br/><br/>
+                        <t t-out="user.name or ''">Mitchell Admin</t> shared the <strong t-out="object.name or ''">document</strong> with you!
+                        <div style="margin: 16px 8px 16px 8px; text-align: center;">
+                            <a t-att-href="object.website_url">
+                                <img t-att-alt="object.name" t-attf-src="{{ ctx.get('base_url') }}/web/image/slide.channel/{{ object.id }}/image_1024" style="height:auto; width:150px; margin: 16px;"/>
```
Comment (std-odoo) [inline addons/website_slides/data/mail_template_data.xml:49]: can you split this line ?

File: addons/website_slides/models/slide_channel.py [hunk 11]
```diff
@@ -233,6 +233,10 @@ def _get_default_enroll_msg(self):
         help="Defines the email your Attendees will receive each time you upload new content.",
         default=lambda self: self.env['ir.model.data']._xmlid_to_res_id('website_slides.slide_template_published'),
         domain=[('model', '=', 'slide.slide')])
+    share_channel_template_id = fields.Many2one(
+        'mail.template', string='Channel Share Template',
+        help="Email template used when sharing a channel",
```
Comment (std-odoo) [inline addons/website_slides/models/slide_channel.py:238]: picky std: use single quote (to be consistent)

File: addons/website_slides/views/website_slides_templates_utils.xml [hunk 12]
```diff
@@ -20,6 +20,14 @@
             social-key="linkedin" aria-label="Share on LinkedIn" title="Share on LinkedIn" target="_blank">
             <i class="fa fa-linkedin fa-fw"/>
         </a>
+        <a t-attf-href="https://wa.me/?text=#{website_share_url}" class="btn border bg-white o_wslides_js_social_share"
+            social-key="whatsapp" aria-label="Share on Whatsapp" title="Share on Whatsapp" target="_blank">
+            <i class="fa fa-whatsapp fa-fw"/>
+        </a>
+        <a t-attf-href="http://pinterest.com/pin/create/button/?url=#{website_share_url}" social-key="pinterest" class="btn border bg-white o_wslides_js_social_share"
```
Comment (std-odoo) [inline addons/website_slides/views/website_slides_templates_utils.xml:27]: can you split this line ?

File: addons/website_slides/models/slide_channel.py [hunk 13]
```diff
@@ -722,6 +726,22 @@ def _remove_membership(self, partner_ids):
         if removed_channel_partner_domain:
             self.env['slide.channel.partner'].sudo().search(removed_channel_partner_domain).unlink()
 
+    def _send_share_email(self, emails):
```
Comment (std-odoo) [inline addons/website_slides/models/slide_channel.py:729]: can you add a docstring to explain what the method does ?

File: addons/website_slides/static/src/xml/website_slides_share.xml [hunk 14]
```diff
@@ -8,37 +8,41 @@
...
                         </button>
                     </div>
                 </div>
             </div>
-            <div t-attf-class="col-12 col-lg-6">
+            <div class="col-12 mt-4">
+                <h5 class="mt-0 mb-2">Share on Social Media</h5>
+                <div class="btn-group" role="group">
+                    <div class="s_share">
+                        <a t-attf-href="https://www.facebook.com/sharer/sharer.php?u=#{widget.slide.websiteShareUrl}" class="btn border bg-white o_wslides_js_social_share" social-key="facebook" aria-label="Share on Facebook" title="Share on Facebook"><i class="fa fa-facebook-square fa-fw"/></a>
```
Comment (std-odoo) [inline addons/website_slides/static/src/xml/website_slides_share.xml:26]: can you split those lines ?

**Review by std-odoo — COMMENTED** [review]
@dht-odoo Validation :)

File: addons/website_slides/controllers/main.py [hunk 15]
```diff
@@ -908,11 +909,13 @@ def slide_preview(self, slide_id):
...
         return slide.is_preview
 
-    @http.route(['/slides/slide/send_share_email'], type='json', auth='user', website=True)
-    def slide_send_share_email(self, slide_id, email, fullscreen=False):
-        slide = request.env['slide.slide'].browse(int(slide_id))
-        result = slide._send_share_email(email, fullscreen)
-        return result
+    @http.route(['/slides/send_share_email'], type='json', auth='user', website=True)
+    def slide_send_share_email(self, emails, res_id, res_model, fullscreen=False):
+        for email in emails.split(','):
```
Comment (std-odoo) [inline addons/website_slides/controllers/main.py:914]: can you check that res_model is either "slide.slide" either "slide.channel" ? (if it's not the case, you can raise a NotFound error)

**Review by std-odoo — COMMENTED** [review]
@dht-odoo Hello :) 

One small last comment, otherwise LGTM, good work :)

File: addons/website_slides/models/slide_channel.py [hunk 16]
```diff
@@ -722,6 +726,23 @@ def _remove_membership(self, partner_ids):
         if removed_channel_partner_domain:
             self.env['slide.channel.partner'].sudo().search(removed_channel_partner_domain).unlink()
 
+    def _send_share_email(self, emails, fullscreen):
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_channel.py:729]: "fullscreen" parameter is not necessary here (since you will split the controller methods).

File: addons/website_slides/controllers/main.py [hunk 17]
```diff
@@ -908,11 +909,15 @@ def slide_preview(self, slide_id):
...
+            if not single_email_re.match(email.strip()):
+                return False
+        if res_model not in ['slide.channel', 'slide.slide']:
+            raise werkzeug.exceptions.NotFound()
+        slide = request.env[res_model].browse(int(res_id))
+        return slide._send_share_email(emails, fullscreen)
```
Comment (awa-odoo) [inline addons/website_slides/controllers/main.py:920]: Not sure I see the point of returning the mail_ids, you could call the method and simply return "True" as the caller does not seem to care.

File: addons/website_slides/static/src/js/slides_embed.js [hunk 18]
```diff
@@ -227,22 +227,32 @@ $(function () {
             var widget = $('.oe_slide_js_share_email');
             var input = widget.find('input');
             var slideID = widget.find('button').data('slide-id');
-            if (input.val() && input[0].checkValidity()) {
```
Comment (awa-odoo) [inline addons/website_slides/static/src/js/slides_embed.js:230]: The runbot is a bit old so I can't test it, but wasn't it done to check if the input email address was correctly formatted?

File: addons/website_slides/static/src/js/slides_embed.js [hunk 19]
```diff
@@ -227,22 +227,32 @@ $(function () {
...
+            if (input.val()) {
                 widget.removeClass('o_has_error').find('.form-control, .custom-select').removeClass('is-invalid');
                 $.ajax({
                     type: "POST",
                     dataType: 'json',
-                    url: '/slides/slide/send_share_email',
```
Comment (awa-odoo) [inline addons/website_slides/static/src/js/slides_embed.js:235]: Since we will re-split the controller methods, will have to see how this can be nicely factorized to avoid too much copy/paste.

File: addons/website_slides/static/src/js/slides_share.js [hunk 20]
```diff
@@ -7,31 +7,64 @@ import { _t } from 'web.core';
...
+        }
+    },
+
     _sendMail: function () {
         var self = this;
+        const channelID = this.$('button').data('channel-id');
         var input = this.$('input');
         var slideID = this.$('button').data('slide-id');
-        if (input.val() && input[0].checkValidity()) {
+        let resID;
```
Comment (awa-odoo) [inline addons/website_slides/static/src/js/slides_share.js:35]: Same here, will have to see how this can be factorize since I would like to keep the routes split.
I think it's possible to keep it nice and with not too much copy/pasting, maybe by simply having a method that takes the route as argument? Will have to see.

File: addons/website_slides/views/website_slides_templates_lesson.xml [hunk 21]
```diff
@@ -210,51 +210,52 @@
...
-                <a t-att-class="'btn btn-light border %s' % ('disabled' if not next_slide else '')"
-                    role="button" t-att-aria-disabled="'disabled' if not next_slide else None"
-                    t-att-href="'/slides/slide/%s' % (slug(next_slide)) if next_slide else '#'">
-                    <span class="d-none d-sm-inline-block">Next</span> <i class="fa fa-chevron-right ml-2"></i>
-                </a>
-            </div>
-            <a class="btn btn-light border ml-2" role="button" t-att-href="'/slides/slide/%s?fullscreen=1' % (slug(slide))">
-                <i class="fa fa-desktop mr-2"/>
-                <span class="d-none d-sm-inline-block">Fullscreen</span>
+    <div class="d-flex my-3 justify-content-end">
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_lesson.xml:213]: Would like a fresh runbot to start checking the xml, seems like a lot of changes but maybe a lot was modified visually.
Hard to see from the base specifications so I will need to compare the screens.

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

Finally gave this one a look, very sorry for the delay.
I did not check the XML code as it's hard to review without having the new screens in front of me to see a bit what was changed exactly.

Could you please take a look at these early comments and most importantly rebase the branch as there seems to be file conflicts?

Thanks & cheers!

File: addons/website_slides/static/src/js/slides_embed.js (hunk 18 above)
Comment (nla-odoo) [reply addons/website_slides/static/src/js/slides_embed.js:230]: Hello @awa-odoo 

Yes, it was done to validate email address, but now we are supporting multiple email addresses hence it is changed.
we can use Email Input with `multiple="True"` attribute. But, The custom 'email' keyboard does not provide a comma key, so users cannot enter multiple email addresses.

File: addons/website_slides/controllers/main.py [hunk 22]
```diff
@@ -17,6 +17,7 @@
 from odoo.exceptions import AccessError, ValidationError, UserError, MissingError
```
Comment (awa-odoo) [inline addons/website_slides/controllers/main.py:17]: Can you rework a bit the commit message to include:

> PURPOSE
> 
> A quick explanation of the goal of these changes in one or 2 sentences
> 
> SPECIFICATIONS
> 
> Your bullet points.

There are also a couple typos:

"vaidation" -> "validation"
"throgh" -> "through"

To apply to both commit messages.

(Side-note: some code editor like the one I use, VSCode, allows you to install a spellchecker addon while typing commit messages, it saves me typos all the time, maybe something to look at).

File: addons/website_slides/views/website_slides_templates_lesson.xml [hunk 23]
```diff
@@ -467,6 +433,24 @@
...
+    <t t-call="website_slides.share_slide_modal">
+        <t t-set="record" t-value="slide"/>
+        <t t-set="include_embed" t-value="bool(slide.embed_code)"/>
+        <t t-set="share_slide_modal_title">Share This Document</t>
+        <t t-set="website_share_url" t-value="slide.website_share_url"/>
+    </t>
+</template>
+
+<!-- Share: social media -->
+<template id='share_slide_modal'>
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_lesson.xml:445]: Looks like we already have a "slide_share_modal" template, we can't have a "share_slide_modal" that looks almost identical.

Can't you re-use the other one?

File: addons/website_slides/views/website_slides_templates_utils.xml [hunk 24]
```diff
@@ -38,17 +48,19 @@
 <template id="slide_share_modal_header">
     <div class="modal-header">
         <h5 class="modal-title" id="slideChannelShareModalLabel">
-            Share
+            <t t-if="share_slide_modal_title" t-out="share_slide_modal_title"/>
+            <t t-else="">Share This Course</t>
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_utils.xml:52]: Side-note, looks like it says "Share This Course" when sharing a slide, not sure if intended or not:

image

File: addons/website_slides/static/src/xml/website_slides_share.xml [hunk 25]
```diff
@@ -8,42 +8,67 @@
...
-                    <button class="btn btn-sm btn-primary o_clipboard_button" style="border-top-end-radius: 4px;border-bottom-end-radius: 4px;" >
-                        <span class="fa fa-clipboard"> Copy Link</span>
+                    <input type="text" class="form-control o_wslides_js_share_link text-center" t-att-value="widget.slide.websiteShareUrl" readonly="readonly" onClick="this.select();" />
+                    <button class="btn btn-sm btn-primary o_clipboard_button" style="border-top-right-radius: 4px;border-bottom-right-radius: 4px;" >
+                        <i class="fa fa-clipboard"/> Copy Link
                     </button>
                 </div>
             </div>
-            <div t-attf-class="col-12 col-lg-6">
+            <div class="col-12 mt-4">
```
Comment (awa-odoo) [inline addons/website_slides/static/src/xml/website_slides_share.xml:20]: It seems like most of this commit is also related to improving the share feature in the website_slides module.

I would re-group all the changes related to sharing in a single commit and maybe make a separate commit for the miscellaneous improvements, such as the small change in "website_slides_forum" and the changes in the review labels.

**Review by awa-odoo — COMMENTED** [review]
Hello @nla-odoo 

Added some extra comments.
Was able to test the feature on a fresh runbot and it's looking good!

I also suggest a small rework on the commits, see underlying comment.

Thanks & cheers!

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-ff
